+nosnoc/+discrete_time_problem/Cls.m

```diff
@@ -0,0 +1,1600 @@
+                            end
+                            if size(model.g_comp_path, 1) > 0
+                                g_comp_path = dcs.g_path_fun(x_ijk, z_ijk, ui, v_global, p);
+                                obj.G.path(ii,jj,kk) = {g_comp_path(:,1)};
+                                obj.H.path(ii,jj,kk) = {g_comp_path(:,2)};
+                            end
+                            if opts.cost_integration
+                                % also integrate the objective
+                                obj.f = obj.f + opts.b_rk(kk)*h*qj;
+                            end
+                        end
+                        if ~opts.right_boundary_point_explicit
+                            x_ijk = obj.w.x(ii,jj,opts.n_s+1);
+                            z_ijk = obj.w.z(ii,jj,opts.n_s+1);
+                            y_gap_ijk = obj.w.y_gap(ii,jj,opts.n_s+1);
+                            
+                            obj.g.dynamics(ii,jj,opts.n_s+1) = {x_ijk - x_ij_end};
+                            obj.g.z(ii,jj,opts.n_s+1) = {dcs.g_z_fun(x_ijk, z_ijk, ui, v_global, p)};
+                            obj.g.y_gap_rbp(ii,jj) = {y_gap_ijk - dcs.f_c_fun(x_ijk)};
+                        else
+                            obj.g.dynamics(ii,jj,opts.n_s+1) = {x_ij_end - obj.w.x(ii,jj,opts.n_s)};
+                        end
+                        if ~opts.g_path_at_stg && opts.g_path_at_fe
+                            obj.g.path(ii,jj) = {dcs.g_path_fun(x_ijk, z_ijk, ui, v_global, p), model.lbg_path, model.ubg_path};
+                        end
+                    end
+                    x_prev = obj.w.x(ii,jj,opts.n_s+rbp);
+                end
+                if ~opts.g_path_at_stg && ~opts.g_path_at_fe
+                    x_i = obj.w.x(ii, opts.N_finite_elements(ii), opts.n_s);
+                    z_i = obj.w.z(ii, opts.N_finite_elements(ii), opts.n_s);
+                    obj.g.path(ii) = {dcs.g_path_fun(x_i, z_i, ui, v_global, p), model.lbg_path, model.ubg_path};
+                end
+
+                % Least Squares Costs
+                % TODO we should convert the refs to params
+                if ~isempty(model.x_ref_val)
+                    obj.f = obj.f + t_stage*dcs.f_lsq_x_fun(obj.w.x(ii,opts.N_finite_elements(ii),opts.n_s),...
+                        model.x_ref_val(:,ii),...
+                        p);
+                end
+                if ~isempty(model.u_ref_val)
+                    obj.f = obj.f + t_stage*dcs.f_lsq_u_fun(obj.w.u(ii),...
+                        model.u_ref_val(:,ii),...
+                        p);
+                end
+                if ~opts.cost_integration
+                    obj.f = obj.f + dcs.f_q_fun(x_ijk, z_ijk, ui, v_global, p);
+                end
+
+                % Clock <Constraints
+                % TODO(@anton) HERE BE DRAGONS. This is by far the worst part of current nosnoc as it requires the discrete problem
```

Comment:
I am not sure i unterstand this comment?

Reply:
Basically this is a leak in the interface between the `model` and the `discrete_time_problem` here. Timefreezing is strictly a model concern however with the current naming scheme it looks like the discrete time problem needs to know whether a time freezing reformulation happened which is not the cleanest thing. In practice this is maybe dealt with renamed flag in `Options` called something like `has_physical_clock_state`, which is more generic.

Reply:
Ahhh  I see, It is not necessary in CLS. removed.